lib/posix/fs.c

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <string.h>
+#include <errno.h>
+#include <posix/unistd.h>
+
+
+struct p_fs_file {
+	struct fs_file_t file;
+	bool used;
+};
+
+struct p_fs_dir {
+	struct fs_dir_t	dir;
+	bool used;
+};
+
+static struct p_fs_dir dir_array[CONFIG_MAX_OPEN_FILES];
```

Comment:
You might want to consider using an union to store both files and directories:

```C
union file_descriptor {
    struct fs_file_t file;
    struct fs_dir_t dir;
};
```

The reason is that you'll want to support `dirfd()` in the future, and you'll need to unify both directory and file descriptors.  This will also simplify some things (e.g. only one mutex), but you'll need another array to store flags/type:

```C
struct file_descriptor_flags {
    u8_t used : 1;
    u8_t is_directory : 1;
    u8_t refs : 6;
};
STATIC_ASSERT(CONFIG_MAX_OPEN_FILES < 1<<6);
static struct file_descriptor_flags fd_flags[CONFIG_MAX_OPEN_FILES];
```

It's also possible to write a function such as: `int find_free_fd(void)` and use it for both open() and opendir().

Another thing to notice is that there's no distinction between a file descriptor (per-process) and file description (per system).  This is fine for an embedded RTOS such as Zephyr, but we need to take into consideration a few other things.  For instance, removing files while they're open is something that should be supported, but no operation should be sent to the file system while they're open; the system must keep track of this and enforce refcounts.

Reply:
I had treated open files and directories separately but this would combine them and have one open fd list which makes sense from unix point of view. Will fix it accordingly.